enhancements/machine-api/machine-deletion-hooks.md

```diff
@@ -0,0 +1,412 @@
+update any troubleshooting guides to check these hook points where possible.
+* Users may add hooks which end up indefinitely blocking removal of Machines.
+We already have a [MachineNotYetDeleted](https://github.com/openshift/machine-api-operator/blob/master/docs/user/Alerts.md#machinenotyetdeleted)
+alert that will warn users when a Machine removal is blocked for an extended
+period. This should continue to function and notify users when a removal has
+been blocked, at which point we expect manual intervention to check the Machine.
+
+## Design Details
+
+### Test Plan
+
+* E2E testing will be added to the cluster-api-actuator-pkg (Machine API E2E suite)
+to ensure that the hooks behave as expected
+
+### Graduation Criteria
+
+This feature is not large enough to require graduation and will be considered GA from
+initial release.
+
+#### Dev Preview -> Tech Preview
+
+This feature will be released straight to GA.
+
+#### Tech Preview -> GA
+
+This feature will be released straight to GA.
+
+#### Removing a deprecated feature
+
+This proposal does not involve removing any deprecated features.
+
+### Upgrade / Downgrade Strategy
+
+As this feature is opt-in, there is no need to account for upgrades.
+
+Any new Hook Implementing Controller will be able to add the hooks at
+any point and they will be observed when the updated Machine controller
+is deployed via the CVO.
+
+On downgrades, HICs may wish to remove the annotations, but they are not
+harmful if left behind.
+
+### Version Skew Strategy
+
+We do not expect any issues with version skew as this is a new feature.
+
+## Implementation History
+
+* This has already been implemented in Cluster API
+
+## Drawbacks
+
+* If misused, this could prevent Machines from being deleted and lead to an increase in support
+cases to determine if it is safe to override the feature.
+
+## Alternatives
+
+### Custom Machine Controller
+Require advanced users to fork and customize.
+This can already be done if someone chooses, so not much of a solution.
+
+### Finalizers
+We define additional finalizers, but this really only implies the deletion lifecycle point.
+A misbehaving controller that accidentally removes finalizers could have undesirable effects.
+
+The proposal today effectively extends the finalizer concept to allow two stages of finalization
+(pre and post drain) in conjunction with the existing Machine finalizer.
+
+If we were to leverage finalizers for this purpose, there are a few different ways this could work,
+as set out below.
+
+#### Machine API waits on all Finalizers
+
+In this scenario, the Machine API would not start draining/removing a Machine until all finalizers
+other than its own one have been removed. This would allow future controllers to manage the removal
+of the Machine without having Machine API have to know anything about their specific implementation.
+However, there could be a scenario where another controller also needs to wait for other finalizers
```

Comment:
Can we add a few examples of these scenarios here?
Multiple controllers operating over machines needing to wait for each others and so introducing state sounds like possibly not the best angle to look at the particular problem being solved and so supporting a design for this might be a flaw.

Since this is proposing to introduce a new public API can we describe at least 2 valid scenarios/use cases that this would solve? By looking at user stories section only Story 1 seems a real need.

Reply:
I'm not sure I can think of a specific example right now, but having suggested this as a design for this problem, I think it's reasonable to think that there may be other cases of this in the future. To me, "waiting for everyone else to go first" is fragile and seems like the kind of thing where if we assume it's fine now, will backfire later.

If we looked at this a slightly different way and use Nodes as an example, on Nodes, if a PVC is used, there are PVC finalizers added that prevent the Node removal until the PVCs have been detached. If we blocked draining the Node waiting for other finalizers to go, this would create a deadlock as the PVC needs the pods to be removed before it can remove the finalizer it created.

I know that doesn't affect this specific situation but if we do go down this route, it would open us up to this kind of dependency in the future